node/src/p2p/swarm_manager.rs

```diff
@@ -0,0 +1,703 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::LazyLock;
+use std::time::Duration;
+
+use futures::StreamExt;
+use libp2p::core::transport::ListenerId;
+use libp2p::identity::Keypair;
+use libp2p::kad::{QueryInfo, RecordKey};
+use libp2p::multiaddr::{Multiaddr, Protocol};
+use libp2p::swarm::dial_opts::{DialOpts, PeerCondition};
+use libp2p::swarm::{ConnectionId, DialError, NetworkBehaviour, NetworkInfo, Swarm, SwarmEvent};
+use libp2p::{autonat, identify, kad, ping, PeerId};
+use lumina_utils::time::Interval;
+use multihash_codetable::{Code, MultihashDigest};
+use tokio::select;
+use tokio::sync::watch;
+use tracing::{debug, error, instrument, trace, warn};
+
+use crate::events::{EventPublisher, NodeEvent};
+use crate::p2p::connection_control;
+use crate::p2p::swarm::new_swarm;
+use crate::p2p::Result;
+use crate::peer_tracker::{Peer, PeerTracker, PeerTrackerInfo, GC_INTERVAL};
+use crate::utils::{celestia_protocol_id, MultiaddrExt};
+
+const MAX_PROTECTED_FULL_NODES: usize = 5;
```

Comment:
I have a slight dislike for the fact that when we say MAX here, we actually mean someting more like DESIRED_AMOUNT. I believe the wording comes from the spec, so it's still probably better to use same terminology as celestia, but could you add a comment here like 
```
// desired number of full node peers, below this threshold node will start searching for new ones
```

Reply:
renamed to FULL_NODES_PROTECT_LIMIT